### ISSUE TITLE
Volume plugins Builder/Cleaner can be type-asserted incorrectly

### ISSUE THREAD

**Issue by thockin** [GoogleCloudPlatform/kubernetes#9139]
For many of the volume plugins, we do something like:

```
func NewBuilder(...) volume.Builder {
  return &whatever {
      name: ...
      type: ...
      path: ...
  }
}

func NewCleaner(...) volume.Cleaner {
  return &whatever {
      name: ...
  }
}
```

Specifically we return the same type but only partially filled in (since we only have partial info at Cleaner time).  Unfortunately, Go will let you type-assert a Cleaner to a Builder in this case.  But calling Builder methods WILL NOT WORK.

Instead, we should have distinct types for whateverBuilder and whateverCleaner that embed a common core which implements volume.Plugin.

@markturansky

**Comment by jiangyaoguo** [comment]
@thockin @markturansky Has anyone been working on this issue?  If nobody, I volunteer to fix this.

**Comment by markturansky** [comment]
@jiangyaoguo We've not started this yet.  Feel free to take a shot at it :)  Between Builders, Cleaners, and now Recyclers, I think the pattern is pretty clear.

I'd start with just 1 in a PR and get some feedback.  Repeat with the rest of the plugins.  I'm happy to review your work.

**Comment by pmorie** [comment]
Likewise, feel free to tag me in for code review.

**Comment by swagiaal** [comment]
Can I request that we hold off on this refactoring until #6234 and its upcoming ilk are merged ? #6234 already contains encapsulation of common core which might make this work a little easier.
@jiangyaoguo sorry I should have made a comment earlier but WDYT ?

**Comment by pmorie** [comment]
@swagiaal We will rebase onto #6234 after it goes in.
